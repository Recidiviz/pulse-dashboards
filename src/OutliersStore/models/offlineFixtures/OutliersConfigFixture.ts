// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { OutliersConfig } from "../OutliersConfig";
import { ADVERSE_METRIC_IDS } from "./constants";

export const OutliersConfigFixture: OutliersConfig = {
  supervisionOfficerLabel: "officer",
  supervisionDistrictLabel: "region",
  supervisionDistrictManagerLabel: "district manager",
  supervisionJiiLabel: "client",
  supervisionSupervisorLabel: "supervisor",
  supervisionUnitLabel: "team",
  learnMoreUrl: "https://recidiviz.org",
  metrics: [
    {
      name: ADVERSE_METRIC_IDS.enum.absconsions_bench_warrants,
      bodyDisplayName: "absconsion rate",
      titleDisplayName: "Absconsion Rate",
      eventName: "absconsions",
      outcomeType: "ADVERSE",
    },
    {
      name: ADVERSE_METRIC_IDS.enum.incarceration_starts,
      bodyDisplayName: "incarceration rate",
      titleDisplayName: "Incarceration Rate",
      eventName: "incarcerations",
      outcomeType: "ADVERSE",
    },
    {
      name: ADVERSE_METRIC_IDS.enum.incarceration_starts_technical_violation,
      bodyDisplayName: "technical incarceration rate",
      titleDisplayName: "Technical Incarceration Rate",
      eventName: "technical incarcerations",
      outcomeType: "ADVERSE",
    },
  ],
  clientEvents: [
    {
      displayName: "Violations",
      name: "violations",
    },
    {
      displayName: "Sanctions",
      name: "violation_responses",
    },
  ],
};
