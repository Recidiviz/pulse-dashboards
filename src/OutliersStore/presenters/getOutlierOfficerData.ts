// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { Optional } from "utility-types";

import { MetricConfig } from "../models/MetricConfig";
import { SupervisionOfficer } from "../models/SupervisionOfficer";
import { OutliersSupervisionStore } from "../stores/OutliersSupervisionStore";
import { OutlierOfficerData } from "./types";

/**
 * Collects all of the officer data, modeling relationships between them with nested objects,
 * and verifies that all of the related objects actually exist. It throws an error rather than
 * returning a partial result, to guarantee return values are fully hydrated.
 */
export function getOutlierOfficerData(
  officerData: SupervisionOfficer,
  supervisionStore: OutliersSupervisionStore
): OutlierOfficerData {
  return {
    ...officerData,
    outlierMetrics: officerData.outlierMetrics.map((metric) => {
      // verify that the related objects we need are actually present;
      // specifically, the metric configs for this officer and the benchmarks
      // for their caseload type
      const metricConfig = supervisionStore.metricConfigsById?.get(
        metric.metricId
      );
      if (!metricConfig) {
        throw new Error(`Missing metric configuration for ${metric.metricId}`);
      }

      const metricBenchmarks = metricConfig.metricBenchmarksByCaseloadType;
      if (!metricBenchmarks) {
        throw new Error(`Missing metric benchmark data for ${metric.metricId}`);
      }

      const caseloadType = officerData?.caseloadType || "ALL";
      const benchmark = metricBenchmarks.get(caseloadType);
      if (!benchmark) {
        throw new Error(
          `Missing metric benchmark data for caseload type ${caseloadType} for ${metric.metricId}`
        );
      }

      const currentPeriodData =
        metric.statusesOverTime[metric.statusesOverTime.length - 1];
      const currentPeriodTarget =
        benchmark.benchmarks[benchmark.benchmarks.length - 1].target;

      // current officer's rate is duplicated in the benchmark values, so we need to remove it
      const filteredBenchmark = { ...benchmark, currentPeriodTarget };
      // making a shallow copy of the array so we don't mutate the one from the datastore
      filteredBenchmark.latestPeriodValues = [
        ...filteredBenchmark.latestPeriodValues,
      ];
      const matchingIndex = filteredBenchmark.latestPeriodValues.findIndex(
        (v) => v.value === currentPeriodData.metricRate
      );
      // in practice we always expect a match,
      // but if we miss we don't want to arbitrarily delete the last element
      if (matchingIndex > -1) {
        filteredBenchmark.latestPeriodValues.splice(matchingIndex, 1);
      }

      const config: Optional<MetricConfig, "metricBenchmarksByCaseloadType"> = {
        ...metricConfig,
      };
      delete config.metricBenchmarksByCaseloadType;

      return {
        ...metric,
        benchmark: filteredBenchmark,
        config,
        currentPeriodData,
      };
    }),
  };
}
