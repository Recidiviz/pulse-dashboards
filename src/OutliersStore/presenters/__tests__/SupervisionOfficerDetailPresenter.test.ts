// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { flowResult } from "mobx";

import { RootStore } from "../../../RootStore";
import { OutliersOfflineAPIClient } from "../../api/OutliersOfflineAPIClient";
import { OutliersConfigFixture } from "../../models/offlineFixtures/OutliersConfigFixture";
import { supervisionOfficerFixture } from "../../models/offlineFixtures/SupervisionOfficerFixture";
import { supervisionOfficerSupervisorsFixture } from "../../models/offlineFixtures/SupervisionOfficerSupervisor";
import { OutliersSupervisionStore } from "../../stores/OutliersSupervisionStore";
import { getOutlierOfficerData } from "../getOutlierOfficerData";
import { SupervisionOfficerDetailPresenter } from "../SupervisionOfficerDetailPresenter";

let store: OutliersSupervisionStore;

jest.mock("../getOutlierOfficerData", () => {
  const original = jest.requireActual(
    "../getOutlierOfficerData"
  ).getOutlierOfficerData;
  return {
    getOutlierOfficerData: jest.fn().mockImplementation(original),
  };
});

const getOutlierOfficerDataMock = getOutlierOfficerData as jest.MockedFunction<
  typeof getOutlierOfficerData
>;

beforeEach(() => {
  store = new OutliersSupervisionStore(
    new RootStore().outliersStore,
    OutliersConfigFixture
  );
  store.setOfficerPseudoId(testOfficer.pseudonymizedId);
  store.setMetricId(testMetric.metricId);
});

afterEach(() => {
  jest.restoreAllMocks();
});

const testOfficer = supervisionOfficerFixture[0];
const testMetric = testOfficer.outlierMetrics[0];
const testSupervisor = supervisionOfficerSupervisorsFixture[0];
let presenter: SupervisionOfficerDetailPresenter;

beforeEach(() => {
  presenter = new SupervisionOfficerDetailPresenter(
    store,
    testOfficer.pseudonymizedId
  );
});

describe("with unit data already hydrated", () => {
  beforeEach(async () => {
    await Promise.all([
      flowResult(
        store.hydrateOfficersForSupervisor(testSupervisor.pseudonymizedId)
      ),
      flowResult(store.hydrateSupervisionOfficerSupervisors()),
      flowResult(store.hydrateMetricConfigs()),
    ]);
  });

  test("is immediately hydrated", () => {
    expect(presenter.isHydrated).toBeTrue();
  });

  test("makes no additional API calls", async () => {
    jest.spyOn(OutliersOfflineAPIClient.prototype, "metricBenchmarks");
    jest.spyOn(OutliersOfflineAPIClient.prototype, "supervisionOfficer");
    jest.spyOn(
      OutliersOfflineAPIClient.prototype,
      "supervisionOfficerSupervisors"
    );
    jest.spyOn(
      OutliersOfflineAPIClient.prototype,
      "supervisionOfficerMetricEvents"
    );

    await presenter.hydrate();

    expect(
      store.outliersStore.apiClient.metricBenchmarks
    ).not.toHaveBeenCalled();
    expect(
      store.outliersStore.apiClient.supervisionOfficerSupervisors
    ).not.toHaveBeenCalled();
    expect(
      store.outliersStore.apiClient.supervisionOfficer
    ).not.toHaveBeenCalled();
    expect(
      store.outliersStore.apiClient.supervisionOfficerMetricEvents
    ).not.toHaveBeenCalled();
  });

  test("has outlierOfficerData", () => {
    expect(presenter.outlierOfficerData).toBeDefined();
    expect(presenter.outlierOfficerData).toMatchSnapshot();
  });

  test("has supervisorInfo", () => {
    expect(presenter.supervisorInfo).toBeDefined();
    expect(presenter.supervisorInfo).toEqual(
      supervisionOfficerSupervisorsFixture[0]
    );
  });
});

test("hydration", async () => {
  jest.spyOn(OutliersOfflineAPIClient.prototype, "metricBenchmarks");
  jest.spyOn(OutliersOfflineAPIClient.prototype, "supervisionOfficer");
  jest.spyOn(
    OutliersOfflineAPIClient.prototype,
    "supervisionOfficerSupervisors"
  );
  jest.spyOn(
    OutliersOfflineAPIClient.prototype,
    "supervisionOfficerMetricEvents"
  );

  expect(presenter.isHydrated).toBeFalse();

  await presenter.hydrate();

  expect(presenter.isHydrated).toBeTrue();
  expect(store.outliersStore.apiClient.metricBenchmarks).toHaveBeenCalled();
  expect(store.outliersStore.apiClient.supervisionOfficer).toHaveBeenCalledWith(
    testOfficer.pseudonymizedId
  );
  expect(
    store.outliersStore.apiClient.supervisionOfficerMetricEvents
  ).toHaveBeenCalledWith(testOfficer.pseudonymizedId, testMetric.metricId);
  expect(
    store.outliersStore.apiClient.supervisionOfficerSupervisors
  ).toHaveBeenCalled();
});

test("has outlierOfficerData", async () => {
  await presenter.hydrate();

  expect(presenter.outlierOfficerData).toBeDefined();
  expect(presenter.outlierOfficerData).toMatchSnapshot();
});

test("has supervisorInfo", async () => {
  await presenter.hydrate();

  expect(presenter.supervisorInfo).toBeDefined();
  expect(presenter.supervisorInfo).toEqual(
    supervisionOfficerSupervisorsFixture[0]
  );
});

test("hydration error in dependency", async () => {
  const err = new Error("fake error");
  jest
    .spyOn(OutliersSupervisionStore.prototype, "hydrateMetricConfigs")
    .mockImplementation(() => {
      throw err;
    });

  await presenter.hydrate();
  expect(presenter.error).toEqual(err);
});

test("error assembling metrics data", async () => {
  const err = new Error("oops");
  getOutlierOfficerDataMock.mockImplementation(() => {
    throw err;
  });

  await presenter.hydrate();
  expect(presenter.error).toBeUndefined();

  expect(presenter.outlierOfficerData).toBeUndefined();
  expect(presenter.error).toEqual(err);
});
