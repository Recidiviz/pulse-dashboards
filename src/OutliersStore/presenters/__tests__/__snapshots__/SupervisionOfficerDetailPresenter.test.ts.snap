// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has outlierOfficerData 1`] = `
Object {
  "caseloadType": "SEX_OFFENSE",
  "displayName": "Duke Ellington",
  "district": "Region D1",
  "externalId": "so1",
  "fullName": Object {
    "givenNames": "Duke",
    "surname": "Ellington",
  },
  "outlierMetrics": Array [
    Object {
      "benchmark": Object {
        "benchmarks": Array [
          Object {
            "endDate": 2023-04-01T00:00:00.000Z,
            "target": 0.028739974981108838,
          },
          Object {
            "endDate": 2023-05-01T00:00:00.000Z,
            "target": 0.028739974981108838,
          },
          Object {
            "endDate": 2023-06-01T00:00:00.000Z,
            "target": 0.028739974981108838,
          },
          Object {
            "endDate": 2023-07-01T00:00:00.000Z,
            "target": 0.028739974981108838,
          },
          Object {
            "endDate": 2023-08-01T00:00:00.000Z,
            "target": 0.028739974981108838,
          },
          Object {
            "endDate": 2023-09-01T00:00:00.000Z,
            "target": 0.028739974981108838,
          },
        ],
        "caseloadType": "SEX_OFFENSE",
        "currentPeriodTarget": 0.028739974981108838,
        "latestPeriodValues": Array [
          Object {
            "targetStatus": "FAR",
            "value": 0.06633147564817059,
          },
          Object {
            "targetStatus": "FAR",
            "value": 0.0756829609662537,
          },
          Object {
            "targetStatus": "FAR",
            "value": 0.0843280708509819,
          },
          Object {
            "targetStatus": "FAR",
            "value": 0.11247207457052615,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.02935971685971686,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.03093744702491948,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.03237250554323725,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.03372447565370045,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.03479006815040747,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.04377548572799233,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.05392229280543655,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.05550486618004866,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.06122620145936426,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.0173108845150581,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.018670076726342712,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.019819721980886187,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.022852491860756323,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.02359100310237849,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.023865568196678438,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.024082871470044866,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.02499315256094221,
          },
        ],
        "metricId": "absconsions_bench_warrants",
      },
      "config": Object {
        "bodyDisplayName": "absconsion rate",
        "eventName": "absconsions",
        "eventNameSingular": "absconsion",
        "name": "absconsions_bench_warrants",
        "outcomeType": "ADVERSE",
        "titleDisplayName": "Absconsion Rate",
      },
      "currentPeriodData": Object {
        "endDate": 2023-09-01T00:00:00.000Z,
        "metricRate": 0.10630035918842831,
        "status": "FAR",
      },
      "metricId": "absconsions_bench_warrants",
      "statusesOverTime": Array [
        Object {
          "endDate": 2023-04-01T00:00:00.000Z,
          "metricRate": 0.098509819,
          "status": "FAR",
        },
        Object {
          "endDate": 2023-05-01T00:00:00.000Z,
          "metricRate": 0.094708509819,
          "status": "FAR",
        },
        Object {
          "endDate": 2023-06-01T00:00:00.000Z,
          "metricRate": 0.0843280708509819,
          "status": "FAR",
        },
        Object {
          "endDate": 2023-07-01T00:00:00.000Z,
          "metricRate": 0.1018842831,
          "status": "FAR",
        },
        Object {
          "endDate": 2023-08-01T00:00:00.000Z,
          "metricRate": 0.11247207457052615,
          "status": "FAR",
        },
        Object {
          "endDate": 2023-09-01T00:00:00.000Z,
          "metricRate": 0.10630035918842831,
          "status": "FAR",
        },
      ],
    },
    Object {
      "benchmark": Object {
        "benchmarks": Array [
          Object {
            "endDate": 2023-04-01T00:00:00.000Z,
            "target": 0.21415916840761745,
          },
          Object {
            "endDate": 2023-05-01T00:00:00.000Z,
            "target": 0.21415916840761745,
          },
          Object {
            "endDate": 2023-06-01T00:00:00.000Z,
            "target": 0.21415916840761745,
          },
          Object {
            "endDate": 2023-07-01T00:00:00.000Z,
            "target": 0.21415916840761745,
          },
          Object {
            "endDate": 2023-08-01T00:00:00.000Z,
            "target": 0.21415916840761745,
          },
          Object {
            "endDate": 2023-09-01T00:00:00.000Z,
            "target": 0.21415916840761745,
          },
        ],
        "caseloadType": "SEX_OFFENSE",
        "currentPeriodTarget": 0.21415916840761745,
        "latestPeriodValues": Array [
          Object {
            "targetStatus": "FAR",
            "value": 0.383876345,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.2156891712217462,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.22289320124847334,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.2513774104683196,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.25765638575152044,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.26617355974580686,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.2842457752511487,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.30401465933699817,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.30873334743074643,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.3092029264536003,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.32831614295994244,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.33674410802650345,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.33724475653700453,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.35949406458970506,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.018847464628730763,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.02775243309002433,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.06404070532502851,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.06958013630081494,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.07086690612561887,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.07159670459003531,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.08246723904202441,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.08595921058829072,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.08735192054565034,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.09013087496913326,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.09335038363171355,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.09793399517037832,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.108173789342659,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.11055245941361763,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.12117619160540669,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.14154601861427094,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.14504271806079871,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.14995891536565326,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.19266297176035893,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.19395979098396954,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.1942350332594235,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.19682613049842076,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.20551801801801803,
          },
        ],
        "metricId": "incarceration_starts",
      },
      "config": Object {
        "bodyDisplayName": "incarceration rate",
        "eventName": "incarcerations",
        "eventNameSingular": "incarceration",
        "name": "incarceration_starts",
        "outcomeType": "ADVERSE",
        "titleDisplayName": "Incarceration Rate",
      },
      "currentPeriodData": Object {
        "endDate": 2023-09-01T00:00:00.000Z,
        "metricRate": 0.4331242583488727,
        "status": "FAR",
      },
      "metricId": "incarceration_starts",
      "statusesOverTime": Array [
        Object {
          "endDate": 2023-08-01T00:00:00.000Z,
          "metricRate": 0.023395936157938592,
          "status": "MET",
        },
        Object {
          "endDate": 2023-09-01T00:00:00.000Z,
          "metricRate": 0.4331242583488727,
          "status": "FAR",
        },
      ],
    },
  ],
  "pseudonymizedId": "hashed-so1",
  "supervisorExternalId": "mdavis123",
}
`;

exports[`with unit data already hydrated has outlierOfficerData 1`] = `
Object {
  "caseloadType": "SEX_OFFENSE",
  "displayName": "Duke Ellington",
  "district": "Region D1",
  "externalId": "so1",
  "fullName": Object {
    "givenNames": "Duke",
    "surname": "Ellington",
  },
  "outlierMetrics": Array [
    Object {
      "benchmark": Object {
        "benchmarks": Array [
          Object {
            "endDate": 2023-04-01T00:00:00.000Z,
            "target": 0.028739974981108838,
          },
          Object {
            "endDate": 2023-05-01T00:00:00.000Z,
            "target": 0.028739974981108838,
          },
          Object {
            "endDate": 2023-06-01T00:00:00.000Z,
            "target": 0.028739974981108838,
          },
          Object {
            "endDate": 2023-07-01T00:00:00.000Z,
            "target": 0.028739974981108838,
          },
          Object {
            "endDate": 2023-08-01T00:00:00.000Z,
            "target": 0.028739974981108838,
          },
          Object {
            "endDate": 2023-09-01T00:00:00.000Z,
            "target": 0.028739974981108838,
          },
        ],
        "caseloadType": "SEX_OFFENSE",
        "currentPeriodTarget": 0.028739974981108838,
        "latestPeriodValues": Array [
          Object {
            "targetStatus": "FAR",
            "value": 0.06633147564817059,
          },
          Object {
            "targetStatus": "FAR",
            "value": 0.0756829609662537,
          },
          Object {
            "targetStatus": "FAR",
            "value": 0.0843280708509819,
          },
          Object {
            "targetStatus": "FAR",
            "value": 0.11247207457052615,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.02935971685971686,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.03093744702491948,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.03237250554323725,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.03372447565370045,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.03479006815040747,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.04377548572799233,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.05392229280543655,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.05550486618004866,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.06122620145936426,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.0173108845150581,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.018670076726342712,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.019819721980886187,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.022852491860756323,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.02359100310237849,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.023865568196678438,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.024082871470044866,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.02499315256094221,
          },
        ],
        "metricId": "absconsions_bench_warrants",
      },
      "config": Object {
        "bodyDisplayName": "absconsion rate",
        "eventName": "absconsions",
        "eventNameSingular": "absconsion",
        "name": "absconsions_bench_warrants",
        "outcomeType": "ADVERSE",
        "titleDisplayName": "Absconsion Rate",
      },
      "currentPeriodData": Object {
        "endDate": 2023-09-01T00:00:00.000Z,
        "metricRate": 0.10630035918842831,
        "status": "FAR",
      },
      "metricId": "absconsions_bench_warrants",
      "statusesOverTime": Array [
        Object {
          "endDate": 2023-04-01T00:00:00.000Z,
          "metricRate": 0.098509819,
          "status": "FAR",
        },
        Object {
          "endDate": 2023-05-01T00:00:00.000Z,
          "metricRate": 0.094708509819,
          "status": "FAR",
        },
        Object {
          "endDate": 2023-06-01T00:00:00.000Z,
          "metricRate": 0.0843280708509819,
          "status": "FAR",
        },
        Object {
          "endDate": 2023-07-01T00:00:00.000Z,
          "metricRate": 0.1018842831,
          "status": "FAR",
        },
        Object {
          "endDate": 2023-08-01T00:00:00.000Z,
          "metricRate": 0.11247207457052615,
          "status": "FAR",
        },
        Object {
          "endDate": 2023-09-01T00:00:00.000Z,
          "metricRate": 0.10630035918842831,
          "status": "FAR",
        },
      ],
    },
    Object {
      "benchmark": Object {
        "benchmarks": Array [
          Object {
            "endDate": 2023-04-01T00:00:00.000Z,
            "target": 0.21415916840761745,
          },
          Object {
            "endDate": 2023-05-01T00:00:00.000Z,
            "target": 0.21415916840761745,
          },
          Object {
            "endDate": 2023-06-01T00:00:00.000Z,
            "target": 0.21415916840761745,
          },
          Object {
            "endDate": 2023-07-01T00:00:00.000Z,
            "target": 0.21415916840761745,
          },
          Object {
            "endDate": 2023-08-01T00:00:00.000Z,
            "target": 0.21415916840761745,
          },
          Object {
            "endDate": 2023-09-01T00:00:00.000Z,
            "target": 0.21415916840761745,
          },
        ],
        "caseloadType": "SEX_OFFENSE",
        "currentPeriodTarget": 0.21415916840761745,
        "latestPeriodValues": Array [
          Object {
            "targetStatus": "FAR",
            "value": 0.383876345,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.2156891712217462,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.22289320124847334,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.2513774104683196,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.25765638575152044,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.26617355974580686,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.2842457752511487,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.30401465933699817,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.30873334743074643,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.3092029264536003,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.32831614295994244,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.33674410802650345,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.33724475653700453,
          },
          Object {
            "targetStatus": "NEAR",
            "value": 0.35949406458970506,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.018847464628730763,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.02775243309002433,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.06404070532502851,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.06958013630081494,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.07086690612561887,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.07159670459003531,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.08246723904202441,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.08595921058829072,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.08735192054565034,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.09013087496913326,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.09335038363171355,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.09793399517037832,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.108173789342659,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.11055245941361763,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.12117619160540669,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.14154601861427094,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.14504271806079871,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.14995891536565326,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.19266297176035893,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.19395979098396954,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.1942350332594235,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.19682613049842076,
          },
          Object {
            "targetStatus": "MET",
            "value": 0.20551801801801803,
          },
        ],
        "metricId": "incarceration_starts",
      },
      "config": Object {
        "bodyDisplayName": "incarceration rate",
        "eventName": "incarcerations",
        "eventNameSingular": "incarceration",
        "name": "incarceration_starts",
        "outcomeType": "ADVERSE",
        "titleDisplayName": "Incarceration Rate",
      },
      "currentPeriodData": Object {
        "endDate": 2023-09-01T00:00:00.000Z,
        "metricRate": 0.4331242583488727,
        "status": "FAR",
      },
      "metricId": "incarceration_starts",
      "statusesOverTime": Array [
        Object {
          "endDate": 2023-08-01T00:00:00.000Z,
          "metricRate": 0.023395936157938592,
          "status": "MET",
        },
        Object {
          "endDate": 2023-09-01T00:00:00.000Z,
          "metricRate": 0.4331242583488727,
          "status": "FAR",
        },
      ],
    },
  ],
  "pseudonymizedId": "hashed-so1",
  "supervisorExternalId": "mdavis123",
}
`;
